components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/CORSRequestHandler.java

```diff
@@ -16,69 +16,191 @@
+		} else {
+			subPath = path.substring(selectedApi.getContext().length());
+		}
+		if ("".equals(subPath)) {
+			subPath = "/";
+		}
+		messageContext.setProperty(RESTConstants.REST_SUB_REQUEST_PATH, subPath);
+
+		if (selectedApi.getResources().length > 0) {
+			for (RESTDispatcher dispatcher : RESTUtils.getDispatchers()) {
+				Resource resource = dispatcher.findResource(messageContext, Arrays.asList(selectedApi.getResources()));
+				if (resource != null) {
+					selectedResource = resource;
+					if (Arrays.asList(resource.getMethods()).contains(httpMethod)) {
+						selectedResourceWithVerb = resource;
+						break;
+					}
+				}
+			}
+		}
+		String resourceString =
+				selectedResourceWithVerb != null ? selectedResourceWithVerb.getDispatcherHelper().getString() : null;
+		String resourceCacheKey = APIUtil
+				.getResourceInfoDTOCacheKey(apiContext, apiVersion, resourceString, httpMethod);
+		messageContext.setProperty(APIConstants.API_ELECTED_RESOURCE, resourceString);
+		messageContext.setProperty(APIConstants.API_RESOURCE_CACHE_KEY, resourceCacheKey);
+		setCORSHeaders(messageContext, selectedResourceWithVerb);
+		if (selectedResource != null && selectedResourceWithVerb != null) {
+				if ("inline".equals(inline)) {
+					messageContext.getSequence("_cors_request_handler").mediate(messageContext);
+				}
+				status =  true;
+			}else if (selectedResource != null && selectedResourceWithVerb == null ){
+			if ("OPTIONS".equalsIgnoreCase(httpMethod)) {
+				messageContext.getSequence("_cors_request_handler").mediate(messageContext);
+				Utils.send(messageContext, HttpStatus.SC_OK);
+				status = false;
+			} else {
+				status = true;
+			}
+		}else{
+			status = true;
+		}
+		return status;
+	}
+
+	public boolean handleResponse(MessageContext messageContext) {
+		messageContext.getSequence("_cors_request_handler").mediate(messageContext);
+		return true;
+	}
+
+	/**
+	 * @param messageContext   message context for set cors headers as properties
+	 * @param selectedResource resource according to the request
+	 */
+	public void setCORSHeaders(MessageContext messageContext, Resource selectedResource) {
+		org.apache.axis2.context.MessageContext axis2MC =
+				((Axis2MessageContext) messageContext).getAxis2MessageContext();
+		Map<String, String> headers =
+				(Map) axis2MC.getProperty(org.apache.axis2.context.MessageContext.TRANSPORT_HEADERS);
+		String requestOrigin = headers.get("Origin");
+		messageContext
+				.setProperty(APIConstants.CORSHeaders.ACCESS_CONTROL_ALLOW_ORIGIN, getAllowedOrigins(requestOrigin));
+		String allowedMethods = "";
+		if (selectedResource != null) {
+			for (String method : selectedResource.getMethods()) {
+				allowedMethods += method + ",";
+			}
+			if (!allowedMethods.isEmpty()) {
+				allowedMethods = allowedMethods.substring(0, allowedMethods.length() - 1);
+			}
+		} else {
+			allowedMethods = "GET,POST,DELETE,OPTIONS";
```

Comment:
What are we allowing everything if selectedResource is null? 
We should use Constants.
Shouldn't HEAD be allowed too?